### PR TITLE
feat: Manage Achievement Templates via GraphQL

### PR DESCRIPTION
**Pull request by Jonasdoubleyou**
- Properly separate common and graphql
- Do not use template.stepName being empty as logic, use type === sequential instead
- Provide GraphQL mutations to create and update achievement templates, prohibit updating achievement logic for activated templates
- Provide GraphQL mutations to activate achievements (ensuring they are consistent) and deactivate (ensuring that no user achievement already exists)
- Expose all Achievement Templates through a public Query

resolves https://github.com/corona-school/project-user/issues/1123
https://github.com/corona-school/project-user/issues/1131
https://github.com/corona-school/project-user/issues/1120

### CODE REVIEW

File: common/achievement/template.ts [hunk 1]
```diff
@@ -88,4 +93,165 @@ async function getTemplatesByMetrics(metricsForAction: Metric[]) {
...
+        }
+
+        currentOrder += 1;
+    }
+
+    // Check that stepNames are set for a group with multiple steps
+    if (group.length > 1) {
+        for (const groupTemplate of group) {
+            if (!groupTemplate.stepName) {
+                throw new PrerequisiteError(`For templates of a sequence, every group template must have a stepName`);
```
Comment (Jonasdoubleyou) [inline common/achievement/template.ts:206]: Is this correct? At least in the integration tests this is not followed?

File: integration-tests/15_achievements.ts [hunk 2]
```diff
@@ -575,532 +575,478 @@ function generateLectures(dates: Date[], match: Match, user: User) {
...
-            image: 'Puzzle_01',
-            achievedImage: '',
-            actionName: 'Termin vereinbaren',
-            actionRedirectLink: 'https://calendly.com',
-            actionType: achievement_action_type_enum.Action,
-            condition: 'student_appointment_booked_events > 0',
-            conditionDataAggregations: {
-                student_appointment_booked_events: { metric: 'student_onboarding_appointment_booked', aggregator: 'count' },
-            },
-            isActive: false,
```
Comment (Jonasdoubleyou) [inline integration-tests/15_achievements.ts:618]: Why was this deactivated? If it stays deactivated the groupOrder is not sequential? Don't we assume this? Or does it only have to be monotonically increasing?

File: common/achievement/get.ts [hunk 3]
```diff
@@ -205,21 +204,21 @@ const assembleAchievementData = async (userAchievements: achievements_with_templ
         actionType: currentAchievementTemplate.actionType,
         achievementType: currentAchievementTemplate.type,
         achievementState: state,
-        steps: currentAchievementTemplate.stepName
```
Comment (Jonasdoubleyou) [inline common/achievement/get.ts:208]: Why did this check stepName? What is the semantics of this?

File: common/achievement/template.ts [hunk 4]
```diff
@@ -88,4 +93,165 @@ async function getTemplatesByMetrics(metricsForAction: Metric[]) {
...
+
+    for (const [name, aggregation] of Object.entries(template.conditionDataAggregations as ConditionDataAggregations)) {
+        if (!isMetric(aggregation.metric)) {
+            throw new PrerequisiteError(`Aggregation ${name} uses unknown metric ${aggregation.metric}`);
+        }
+
+        // TODO: How to evaluate the aggregators?
+    }
+
+    // TODO: Does the template.type imply any prerequisities?
```
Comment (Jonasdoubleyou) [inline common/achievement/template.ts:175]: Can you comment on this?

File: common/achievement/template.ts [hunk 5]
```diff
@@ -88,4 +93,165 @@ async function getTemplatesByMetrics(metricsForAction: Metric[]) {
...
+    if (aggregations.length < 1) {
+        throw new PrerequisiteError(`AchievementTemplate needs at least one data aggregation`);
+    }
+
+    for (const [name, aggregation] of Object.entries(template.conditionDataAggregations as ConditionDataAggregations)) {
+        if (!isMetric(aggregation.metric)) {
+            throw new PrerequisiteError(`Aggregation ${name} uses unknown metric ${aggregation.metric}`);
+        }
+
+        // TODO: How to evaluate the aggregators?
```
Comment (Jonasdoubleyou) [inline common/achievement/template.ts:172]: And this?

File: common/achievement/template.ts (hunk 1 above)
Comment (dhenkel92) [reply common/achievement/template.ts:206]: I think it's only correct for sequential achievements. Tiered are in a group as well, but don't have a step name.

File: common/achievement/template.ts [hunk 6]
```diff
@@ -88,4 +93,165 @@ async function getTemplatesByMetrics(metricsForAction: Metric[]) {
...
+        throw new PrerequisiteError(`Cannot deactivate AchievementTemplate as it is already in use`);
+    }
+
+    if (group.length > 1 && group.some((other) => other.groupOrder > template.groupOrder && template.isActive)) {
+        throw new PrerequisiteError(`Cannot deactivate AchievementTemplate with groupOrder ${template.groupOrder} as a following template is still active`);
+    }
+
+    await prisma.achievement_template.update({
+        where: { id },
+        data: { isActive: true },
```
Comment (dhenkel92) [inline common/achievement/template.ts:251]:
```suggestion
        data: { isActive: false },
```

File: common/achievement/template.ts [hunk 7]
```diff
@@ -88,4 +93,165 @@ async function getTemplatesByMetrics(metricsForAction: Metric[]) {
...
+    }
+
+    const hasAchivements = (await prisma.user_achievement.count({ where: { templateId: id } })) > 0;
+    if (hasAchivements) {
+        throw new PrerequisiteError(`Cannot deactivate AchievementTemplate as it is already in use`);
+    }
+
+    if (group.length > 1 && group.some((other) => other.groupOrder > template.groupOrder && template.isActive)) {
+        throw new PrerequisiteError(`Cannot deactivate AchievementTemplate with groupOrder ${template.groupOrder} as a following template is still active`);
+    }
```
Comment (dhenkel92) [inline common/achievement/template.ts:247]: I have no clear opinion if this should be allowed or not.

File: types/custom.d.ts [hunk 8]
```diff
@@ -1,4 +1,4 @@
 declare module '@onlabsorg/swan-js' {
-    declare function parse(condition: any): (context: any) => Promise<boolean>;
+    declare function parse(condition: any): (context: any) => Promise<any>;
```
Comment (dhenkel92) [inline types/custom.d.ts:2]: 👍 Good Catch

File: graphql/achievement_template/mutations.ts [hunk 9]
```diff
@@ -0,0 +1,124 @@
...
+    @Field({ nullable: true })
+    achievedText: string | null;
+    @Field({ nullable: true })
+    achievedImage: string | null;
+    @Field()
+    description: string;
+    @Field()
+    subtitle: string;
+    @Field()
+    stepName: string;
```
Comment (dhenkel92) [inline graphql/achievement_template/mutations.ts:30]: I think this should be optional

File: common/achievement/template.ts (hunk 4 above)
Comment (dhenkel92) [reply common/achievement/template.ts:175]: See slack

File: common/achievement/template.ts (hunk 7 above)
Comment (Jonasdoubleyou) [reply common/achievement/template.ts]: I removed this again, I guess as we only require total order of the group order one could do this

File: graphql/achievement_template/mutations.ts (hunk 9 above)
Comment (Jonasdoubleyou) [reply graphql/achievement_template/mutations.ts:30]: Nope, can be an empty string though :/

**Review by dhenkel92 — APPROVED** [review]
😍

**Pull request merged**